docs/design-proposals/multiple-statefulsets.md

```diff
@@ -0,0 +1,417 @@
+
+So add an extra field to existing may be a better solution.
+
+For example: change `values.yaml` from figure 1 to figure 2, and change the TidbCluster object from figure 3 to figure 4:
+
+Users can create multiple statefulsets in `charts/tidb-cluster/values.yaml`, for example: change one tikv(Figure 1) to more tikvs(Figure 2):
+
+Figure 1:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+```
+
+Figure 2:
+
+``` yaml
+tikv:
+  replicas: 3
+  image: pingcap/tikv:v2.1.0
+  …
+
+tikvs:
+  - name: name-1
+    replicas: 2
+    image: pingcap/tikv:v2.1.0
+    …
+  - name: name-2
+    replicas: 2
+    image: pingcap/tikv:v2.1.0
+    …
+```
+
+The `tikv` section is the main tikv statefulset. `tikvs` is the extra tikv statefulsets. They must include all the attributes.
+
+And each statefulset has a separate ConfigMap named by the statefulset name, the `tikv` statefulset name is: `<clusterName>-tikv`, the `tikvs` statefulset name are `<clusterName>-tikv-name-1` and `<clusterName>-tikv-name-2`.
+
+We must add a name attribute(for example: name-1 or name-2) to the `tikvs` section to distinguish the different statefulsets.
+
+The `tikvs` section can be empty.
+
+The same as TidbCluster object, from Figure 3 to Figure 4:
+
+Figure 3:
+
+``` go
+type TidbClusterSpec struct {
+  TiKV            TiKVSpec            `json:"tikv,omitempty"`
+  …
+}
+
+type TiKVStatus struct {
+  StatefulSet     *apps.StatefulSetStatus     `json:"statefulSet,omitempty"`
+  …
+}
+```
+
+Figure 4:
+
+``` go
+type TidbClusterSpec struct {
+  TiKV         TiKVSpec         `json:"tikv,omitempty"`
+  TiKVs        []TiKVSpec       `json:"tikvs,omitempty"`
+  …
+}
+
+type TiKVSpec struct {
+  Name            string           `json:"name"`
+  Replicas        int32            `json:"replicas"`
+  …
+}
+
+type TiKVStatus struct {
+  StatefulSet     *apps.StatefulSetStatus     `json:"statefulSet,omitempty"`
+  StatefulSets    []apps.StatefulSetStatus    `json:"statefulSets,omitempty"`
+  …
+}
+```
+
+### Create:
+
+The TiDB Operator should maintain multiple statefulset objects in the kube apiserver. The startup scripts are the same as before basically. An exception is PD startup script: the bootstrapping Pod may be the main statefulset first Pod.
+
+The TiDB Operator will create extra Services together with multiple statefulsets:
+
+* For PD: no change
+* For TiKV: no change
+* For TiDB: N+1 Services should be created, N client services and 1 main client service
+
+### Upgrade:
+
+With multiple statefulsets supported, the user can modify the spec in the `values.yaml` then the Operator will range over all the statefulsets and select only one statefulset to upgrade at a time. So the basic progress is the same with older design.
+
+The user can modify all the statefulsets, but we only support upgrade one Pod at a time.
+
+For example, the user can change the images from v2.1.0(Figure 5) to v2.2.0(Figure 6) to upgrade the cluster from v2.1.0 to v2.2.0:
```

Comment:
How do you deal with situations where a user has modified the name of an existing statefulset?

Reply:
Add a new statefulset and keep the old if the user changes the name of an existing one.

Reply:
That would cause the actual running statefulset to be one more than the user configured, so that's fine?

Reply:
I think we should reject that change.

Reply:
what do we need to do to reject it?

Reply:
Its worth noting the difference between what we want to do and what is convenient to do. I think it is not possible to run custom validation logic for a CRD on update. One approach would be to use a phase field to indicate the CRD is an error state.

Reply:
The use change the `values.yaml` file, not the CRD object. The CRD object is updated by us with the help of `helm`.

What we should consider first is how do we modify the CRD if the use change the name in `values.yaml`.

Reply:
We can use `admission webhooks ` to validate the CRD object.

https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#admission-webhooks

Reply:
this is added to the document.

Reply:
Isn't that just for CRD creation? We would also need update.

Reply:
[Update](https://github.com/kubernetes/kubernetes/blob/37555e6d24c2f951c40660ea59a80fa251982005/pkg/apis/admissionregistration/types.go#L259) is supported.